### PR TITLE
[ROCm][perf] Use workspace manager for sparse indexer allocations

### PR DESCRIPTION
**Pull request by tuukkjs**
## Purpose

Replace dynamic per-call allocations in the ROCm sparse attention indexer with workspace manager allocations.

The sparse attention indexer uses temporary buffers in both prefill and decode: `k_fp8` / `k_scale` in prefill and decode logits buffers in `rocm_fp8_paged_mqa_logits`. These buffers are runtime scratch space, and their sizes vary with request shape, batch size, and speculative decoding state. They are a good fit for `WorkspaceManager`, which is already used elsewhere in vLLM to reserve and reuse bounded scratch buffers instead of relying on repeated hot-path PyTorch allocator calls.

This PR is primarily an allocator robustness and consistency change:

- It avoids repeated dynamic `torch.empty` / `torch.full` allocations in the ROCm sparse indexer hot path.
- It makes the ROCm sparse prefill behavior match the existing CUDA sparse indexer pattern, where temporary buffers are workspace-managed.
- It lets profiling reserve the required scratch size up front, so later execution reuses the profiled workspace instead of growing allocator state unexpectedly.

Earlier validation of the stale predecessor PR showed plain-image failures under sustained MTP load, including memory access faults and `HSA_STATUS_ERROR_OUT_OF_RESOURCES`. On the current nightly stack used for the tests, that exact MTP=2 failure no longer reproduces reliably in the plain image. Therefore this PR can not be said to fix MTP=2 crash that occured eariler. The argument for PR is that sparse-indexer scratch allocations should be workspace-managed on ROCm for the same reason they are workspace-managed elsewhere: bounded scratch reuse, less allocator churn, and lower risk of fragmentation or late hot-path allocation surprises.

This continues the work of stale PR #38704 by @gronsti-amd, rebased onto current `main`. AI assistance was used for this PR.

## Solution

- **Prefill path:** Replace per-chunk `torch.empty` allocations for `k_fp8` and `k_scale` with a single `workspace_manager.get_simultaneous()` allocation, slicing the workspace views per chunk.
- **Decode path:** Allocate the temporary decode logits buffer from `WorkspaceManager` and pass it to the existing AITER paged-MQA logits kernels.
- **Memory profiling path:** Reserve workspace for the prefill and decode buffer sizes in the real implementation's early-exit path, while keeping the fake implementation compatible with FakeTensor mode during `torch.compile`.

## Test Plan

**Environment:**

- Image: `vllm/vllm-openai-rocm:nightly-f6490a28413526adefae8dd7af32906d060e0436`
- Model: `deepseek-ai/DeepSeek-V3.2`
- Hardware: 8x MI355X, TP=8
- Attention backend: `ROCM_AITER_MLA_SPARSE`

**Four experiments were run for validation:**

1. **No MTP (baseline):** Verifies the workspace manager changes do not regress the non-speculative decoding path.
2. **MTP=1 (1 speculative token):** Validates workspace-managed allocations under MTP with a single draft token.
3. **MTP=2 (2 speculative tokens):** Validates workspace-managed allocations under MTP with two draft tokens.
4. **MTP=3 (3 speculative tokens):** Validates workspace-managed allocations under MTP with three draft tokens.

**Server command:**

```bash
VLLM_ROCM_USE_AITER=1 vllm serve deepseek-ai/DeepSeek-V3.2 \
    --host 0.0.0.0 \
    --port 8000 \
    --seed 4321 \
    --dtype auto \
    --gpu-memory-utilization 0.75 \
    --max-num-seqs 256 \
    --max-model-len 16384 \
    --max-num-batched-tokens 4096 \
    --no-enable-prefix-caching \
    --no-async-scheduling \
    --block-size 64 \
    --attention-backend ROCM_AITER_MLA_SPARSE \
    --compilation-config '{"custom_ops": ["-rms_norm", "-silu_and_mul", "-quant_fp8"]}' \
    --tensor-parallel-size 8 \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": N}'
```

(The `--speculative-config` argument is omitted for the no-MTP baseline, and `N` is 1, 2, or 3 for the MTP experiments.)

**Benchmark command:**

```bash
vllm bench serve \
    --backend vllm \
    --dataset_name random \
    --ignore_eos \
    --model deepseek-ai/DeepSeek-V3.2 \
    --ready_check_timeout_sec 3600 \
    --percentile_metrics ttft,tpot,itl,e2el \
    --random_input_len 1024 \
    --random_output_len 8192 \
    --max_concurrency 128 \
    --temperature 0 \
    --seed 3 \
    --num_warmups 128 \
    --num_prompts 384 \
    --port 8000
```

## Test Results

| Metric | No MTP | MTP=1 | MTP=2 | MTP=3 |
|--------|-------:|------:|------:|------:|
| Successful requests | 384 | 384 | 384 | 384 |
| Failed requests | 0 | 0 | 0 | 0 |
| Benchmark duration (s) | 1436.04 | 1171.12 | 1154.04 | 1047.42 |
| Total input tokens | 393,216 | 393,216 | 393,216 | 393,216 |
| Total generated tokens | 3,145,728 | 3,145,728 | 3,145,728 | 3,145,728 |
| Output token throughput (tok/s) | 2190.56 | 2686.09 | 2725.84 | 3003.31 |
| Total token throughput (tok/s) | 2464.38 | 3021.85 | 3066.57 | 3378.73 |
| Mean TTFT (ms) | 1889.58 | 1774.80 | 1834.01 | 1928.87 |
| P99 TTFT (ms) | 8198.26 | 8650.36 | 9254.22 | 9517.48 |
| Mean TPOT (ms) | 58.18 | 45.88 | 42.11 | 38.67 |
| Mean ITL (ms) | 58.18 | 89.30 | 118.01 | 139.91 |
| Mean E2EL (ms) | 478461.72 | 377606.14 | 346750.78 | 318686.48 |
| Acceptance rate (%) | - | 94.63 | 90.14 | 87.29 |
| Mean acceptance length | - | 1.95 | 2.80 | 3.62 |

### CODE REVIEW

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -511,6 +503,37 @@ def rocm_aiter_sparse_attn_indexer(
...
+        # workspace manager operations on the locked real workspace
+        # would corrupt PyTorch's dispatch state.
+        workspace_manager = current_workspace_manager()
+
+        # Prefill k_fp8 and k_scale buffers, used by
+        # rocm_aiter_sparse_attn_indexer's prefill path
+        workspace_manager.get_simultaneous(
+            ((total_seq_lens, head_dim), fp8_dtype),
+            ((total_seq_lens, 4), torch.uint8),
+        )
```
[the remark this answers is not in the document]
Comment (tuukkjs) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:660]: `total_seq_lens` is not batch-specific. It is `max_model_len * 40`, a fixed upper bound that is identical during profiling and runtime. The workspace manager sees the maximum possible size during profiling and pre-allocates accordingly. No under-reservation is possible.

The CUDA path (`sparse_attn_indexer.py:110-117`) uses the exact same value in the exact same way:

```python
# CUDA profiling path — identical pattern
values_spec, scales_spec = _gather_workspace_shapes(
    total_seq_lens, head_dim, fp8_dtype, use_fp4_cache
)
current_workspace_manager().get_simultaneous(values_spec, scales_spec, ...)
```

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -552,17 +575,15 @@ def rocm_aiter_sparse_attn_indexer(
     if has_prefill:
         prefill_metadata = layer_attn_metadata.prefill
         assert prefill_metadata is not None
+
+        workspace_manager = current_workspace_manager()
+        k_fp8_full, k_scale_full = workspace_manager.get_simultaneous(
+            ((total_seq_lens, head_dim), fp8_dtype),
+            ((total_seq_lens, 4), torch.uint8),
+        )
```
[the remark this answers is not in the document]
Comment (tuukkjs) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:745]: Since `total_seq_lens` is already `max_prefill_buffer_size` (the maximum allowed chunk size from `split_indexer_prefill_chunks`), it is effectively already the max chunk size. The chunking logic in `indexer.py` guarantees that no single chunk's `total_seq_lens` exceeds this value.

Furthermore, the workspace manager's buffer only grows and is never freed during execution. Even if we requested a smaller size, the buffer would already be at this size from the profiling run. There is no memory savings to be had.

The CUDA path (`sparse_attn_indexer.py:186-196`) uses the identical allocation pattern:

```python
# CUDA runtime prefill path — identical pattern
values_spec, scales_spec = _gather_workspace_shapes(
    total_seq_lens, head_dim, fp8_dtype, use_fp4_cache
)
k_quant_full, k_scale_full = workspace_manager.get_simultaneous(
    values_spec, scales_spec,
)
for chunk in prefill_metadata.chunks:
    k_quant = k_quant_full[: chunk.total_seq_lens]
    k_scale = k_scale_full[: chunk.total_seq_lens]
```

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -552,17 +575,15 @@ def rocm_aiter_sparse_attn_indexer(
...
-                device=k.device,
-                dtype=fp8_dtype,
-            )
-            k_scale = torch.empty(
-                [chunk.total_seq_lens, 4],
-                device=k.device,
-                dtype=torch.uint8,
-            )
+            k_fp8 = k_fp8_full[: chunk.total_seq_lens]
+            k_scale = k_scale_full[: chunk.total_seq_lens]
```
[the remark this answers is not in the document]
Comment (tuukkjs) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:748]: See above comment. The allocation size (`total_seq_lens = max_prefill_buffer_size`) is already the max chunk size. The pattern is intentional and matches the CUDA path exactly.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 4]
```diff
@@ -511,6 +503,37 @@ def rocm_aiter_sparse_attn_indexer(
...
+        workspace_manager.get_simultaneous(
+            ((heads, hidden_states.shape[0], max_model_len), torch.float32),
+        )
+
+        # Transient logits tensor peak memory, produced by
+        # rocm_fp8_mqa_logits (prefill) and rocm_fp8_paged_mqa_logits
+        # (decode). Prefill logits are bounded by
+        # VLLM_SPARSE_INDEXER_MAX_LOGITS_MB via chunking in
+        # split_indexer_prefill_chunks; decode logits are smaller.
+        max_logits_elems = envs.VLLM_SPARSE_INDEXER_MAX_LOGITS_MB * 1024 * 1024
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:680]: We should not call `envs.VLLM_SPARSE_INDEXER_MAX_LOGITS_MB` in the hotpath. It is extremely slow.

Given that this is also not traced by torch compile or cuda graph, the CPU over head is noticeable and we have many layers per step and the overhead piles up. 

https://github.com/vllm-project/vllm/issues/17067

Comment (tuukkjs) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:680]: In my understanding `envs.VLLM_SPARSE_INDEXER_MAX_LOGITS_MB` call is not here in the hot path. It is inside the `if not isinstance(attn_metadata, dict)` branch, i.e. the profiling / dummy-run early exit, not the normal prefill or decode execution path. Therefore, it should not in my understanding affect perf. I also ran an experiment which showed that all `envs.VLLM_SPARSE_INDEXER_MAX_LOGITS_MB` calls happened on server startup before server was ready to receive requests. Furthermore, it seeems CUDA sparse-indexer implementation [does the same env lookup in its profiling branch](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/sparse_attn_indexer.py#L108-L123).

**Comment by tuukkjs** [comment]
@tjtanaa, could you have a look at this PR? I reran the tests and updated the PR description. The argument for the PR is no longer about fixing MTP=2 but following a better pattern for sparse indexer allocations. CUDA side already uses the same `WorkspaceManager` pattern for sparse indexer scratch buffers. Pre-commit checks for the file edited should pass. I'd like to resolve this PR either way - merge or close - to not keep it hanging.

**Comment by tjtanaa** [comment]
@tuukkjs overall it looks good. Could you help to get the accuracy score for the model that can validate this changes? e.g. deepseekv3.2? Thanks.

**Comment by tuukkjs** [comment]
> @tuukkjs overall it looks good. Could you help to get the accuracy score for the model that can validate this changes? e.g. deepseekv3.2? Thanks.

@tjtanaa here are the results.  I ran an accuracy comparison for `deepseek-ai/DeepSeek-V3.2` on `gsm8k` (5-shot) on the same `MI350X` node and the same nightly image `vllm/vllm-openai-rocm:nightly-28ee78af543c563a2fbf78829a7688120e4e4eb5`. The other run included changes in PR other used just the plain image.

| Mode | Plain image (`strict-match`) | PR overlay (`strict-match`) | Delta |
|------|------------------------------:|----------------------------:|------:|
| no MTP | 0.6740 | 0.6854 | +0.0114 |
| MTP=2 | 0.5686 | 0.5656 | -0.0030 |

So, it seems there is no there is no accuracy effect with this PR.

**Comment by tuukkjs** [comment]
@tjtanaa, is accuracy testing sufficient and this good to go?

**Comment by tuukkjs** [comment]
I merged main, resolved conflicts and reran the same validations on MI350X node using the latest nightly image:

`vllm/vllm-openai-rocm:nightly-b50646e5effd7cb5884cd96fdff4c53c18521198`

| Setup | `exact_match,strict-match` | `exact_match,flexible-extract` |
|---|---:|---:|
| Plain image, no MTP | `0.9530` | `0.9522` |
| PR overlay, no MTP | `0.9469` | `0.9469` |
| Plain image, MTP=2 | `0.9568` | `0.9560` |
| PR overlay, MTP=2 | `0.9553` | `0.9553` |

Deltas vs plain image (that is, this PR not included):

- No MTP: `-0.0061` strict-match, `-0.0053` flexible-extract
- MTP=2: `-0.0015` strict-match, `-0.0007` flexible-extract

No clear accuracy regression observed; the latest run shows small negative deltas within roughly one standard error

**Comment by tjtanaa** [comment]
@tuukkjs  please use the following evaluation.

For DeepSeek V3.2 and GLM-5.1-FP8, please use this lmeval command. High concurrency can pick up topk logic issue in the indexer. Large num shot (need to be at least 20) to trigger the sparse indexer.

```
MODEL=deepseek-ai/DeepSeek-V3.2
lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8894/v1/completions,num_concurrent=256,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000,trust_remote_code=True --batch_size auto --tasks gsm8k --num_fewshot 20 \
  --output_path ./results_dsv32_numshot20_cc256 \
  --log_samples \
| tee lmeval_dsv32_numshot20_cc256.log
```

both strict and flexible match must be `0.95xx`

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 5]
```diff
@@ -647,6 +645,43 @@ def rocm_aiter_sparse_attn_indexer(
...
+                    (q_fp8.shape[1], hidden_states.shape[0], max_model_len),
+                    torch.float32,
+                ),
+            )
+        # Transient logits tensor peak memory, produced by
+        # rocm_fp8_mqa_logits (prefill) and rocm_fp8_paged_mqa_logits
+        # (decode). Prefill logits are bounded by
+        # VLLM_SPARSE_INDEXER_MAX_LOGITS_MB via chunking in
+        # split_indexer_prefill_chunks; decode logits are smaller.
+        max_logits_elems = envs.VLLM_SPARSE_INDEXER_MAX_LOGITS_MB * 1024 * 1024
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:680]: @tuukkjs Find a way to cache this, we should not use `envs.xxx` in hotpath

As mentioned in this https://github.com/vllm-project/vllm/issues/17067

Comment (tuukkjs) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:680]: Thanks for the comment. Could you elaborate why this is in the hot path (see my previous comment above)? Cuda side seems to do it in same way.

Comment (tjtanaa) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:680]: Thanks for raising this up. Yes. This is only invoked during profiling for the fake function path.

**Review by tjtanaa — APPROVED** [review]

**Comment by tuukkjs** [comment]
> @tuukkjs please use the following evaluation.
> 
> For DeepSeek V3.2 and GLM-5.1-FP8, please use this lmeval command. High concurrency can pick up topk logic issue in the indexer. Large num shot (need to be at least 20) to trigger the sparse indexer.
> 
> ```
> MODEL=deepseek-ai/DeepSeek-V3.2
> lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8894/v1/completions,num_concurrent=256,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000,trust_remote_code=True --batch_size auto --tasks gsm8k --num_fewshot 20 \
>   --output_path ./results_dsv32_numshot20_cc256 \
>   --log_samples \
> | tee lmeval_dsv32_numshot20_cc256.log
> ```
> 
> both strict and flexible match must be `0.95xx`

Thanks, I ran the suggested evals. Here are the results (using vllm/vllm-openai-rocm:nightly-3fd9d2d35714e80b4cb3fcd3c408a0398fa2525f as baseline):

| Model | Build | flexible-extract | strict-match |
|---|---:|---:|---:|
| `deepseek-ai/DeepSeek-V3.2` | baseline nightly | `0.9522` | `0.9522` |
| `deepseek-ai/DeepSeek-V3.2` | PR overlay | `0.9530` | `0.9538` |
| `zai-org/GLM-5.1-FP8` | baseline nightly | `0.9507` | `0.9515` |
| `zai-org/GLM-5.1-FP8` | PR overlay | `0.9575` | `0.9583` |

Both strict and flexible matches are in the requested `0.95xx` range for the PR overlay on both models. For GLM-5.1-FP8, the default lm-eval tokenizer path failed before issuing requests with `Tokenizer class TokenizersBackend does not exist`, so I reran with `tokenizer_backend=None,tokenized_requests=False` while keeping the same concurrency, shot count, generation limit, and GSM8K task. DeepSeek used the command as requested.

**Comment by AndreasKaratzas** [comment]
The issue in unrelated so it can be force merged.

**Comment by tuukkjs** [comment]
Yep, as noted, the CI failures look unrelated
